src/cockpit/samples.py

```diff
@@ -101,14 +103,23 @@ class CPUTemperatureSampler(Sampler):
         SampleDescription('cpu.temperature', 'celsius', 'instant', True),
     ]
 
-    def detect_cpu_sensors(self, hwmonid: int, name: str) -> None:
-        for index in range(1, 2 ** 32):
-            sensor_path = f'/sys/class/hwmon/hwmon{hwmonid}/temp{index}_input'
-            if not os.path.exists(sensor_path):
-                break
+    def detect_cpu_sensors(self, name: str, hwmon_name: str, dir_fd: int) -> None:
+        for sensor in os.listdir(dir_fd):
+            if not sensor.endswith('_input'):
+                continue
+
+            # Try opening the corresponding label
+            label = None
+            try:
+                fd = os.open(sensor.replace('_input', '_label'), os.O_RDONLY, dir_fd=dir_fd)
+                try:
+                    label = os.read(fd, 1024)
+                    label = label.decode().strip()
+                finally:
+                    os.close(fd)
+            except FileNotFoundError:
+                pass
 
-            with open(f'/sys/class/hwmon/hwmon{hwmonid}/temp{index}_label') as f:
-                label = f.read().strip()
             if label:
```

Comment:
I know this logic is pre-existing, but it seems wrong.  If labels are not used on arm, and we should accept all samplers if `name == 'cpu_thermal'`, then why do we even bother trying to open the label file?

In general, the logic here seems to be:
 - `name == 'cpu_thermal'`: accept all inputs which don't have a label file (but the comment suggests that in this case, that will be all inputs — so why check?)
 - `name == 'atk0110'`: accept results only if label is `'CPU Temperature'`
 - `name` is absolutely anything else?  accept everything except `label == 'Tctl'`?

I had an idea to (partially) lift this logic to the calling function, which is where we sample `name`, and could make the decision about what check to use just once, instead of repeatedly.  The exact nature of the checks being performed makes that inelegant, though.

I also dislike how so many separate parameters (`name`, hwmon_name`, `dirfd` and implicit access to the global `HWMON_PATH`) are inputs to this function.

It would be interesting if this function returned something more like a list of sensor input basenames and their labels, if applicable, else `None`.  If you wanted to get really fancy, you could add a `read_label` kwarg.  You could also pass in a predicate function to decide if a particular label is to be considered valid or not (or `None` to mean "all labels are valid, so don't bother reading it").

This is certainly some kind of over-engineering.  This problem well and truly drives me crazy.  Every time I look at this code, I hate it a bit more :)